pyccel/parser/semantic.py

```diff
@@ -960,104 +884,35 @@ def _infere_type(self, expr, **settings):
                 d_var['allocatable'] = d_var['rank']>0
                 d_var['is_pointer' ] = False
 
-            elif name in [
-                    'Abs',
-                    'sin',
-                    'cos',
-                    'exp',
-                    'log',
-                    'csc',
-                    'cos',
-                    'sec',
-                    'tan',
-                    'cot',
-                    'asin',
-                    'acsc',
-                    'acos',
-                    'asec',
-                    'atan',
-                    'acot',
-                    'sinh',
-                    'cosh',
-                    'tanh',
-                    'atan2',
-                    ]:
+            elif isinstance(expr, NumpyUfuncBase):
                 d_var = self._infere_type(expr.args[0], **settings)
-                d_var['datatype'] = sp_dtype(expr)
+                d_var['datatype'   ] = expr.dtype
+                d_var['is_pointer' ] = False
+                d_var['precision'  ] = expr.precision
+
+            elif isinstance(expr, MathFunctionBase):
+                d_var['datatype' ] = expr.dtype
+                d_var['precision'] = expr.precision
+                d_var['rank'     ] = expr.rank
+                d_var['shape'    ] = expr.shape
```

Comment:
I am surprised that these 2 cases are not handled in the same way. Does NumpyUfuncBase not have a shape and rank? I would have expected this to be necessary to handle broadcasting rules?

Reply:
it should be handled the same way

Reply:
@saidctb Do you mean that you could change the lines
```
d_var = self._infere_type(expr.args[0], **settings)
d_var['datatype'] = sp_dtype(expr)
```
to
```
d_var['datatype' ] = expr.dtype
```
and the code would still work correctly?

Reply:
yes I suggest to leave that for another pull request, when we deal with type inference method

Reply:
OK.